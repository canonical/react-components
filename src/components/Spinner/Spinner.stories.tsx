import { Meta, StoryObj } from "@storybook/react";

import Spinner from "./Spinner";

const meta: Meta<typeof Spinner> = {
  component: Spinner,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Spinner>;

export const Default: Story = {
  name: "Default",
};

export const Text: Story = {
  name: "Text",

  args: {
    text: "Loading...",
  },
};

export const Assertive: Story = {
  name: "Assertive",

  args: {
    ariaLive: "assertive",
  },
};
