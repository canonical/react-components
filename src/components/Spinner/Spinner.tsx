import React, { HTMLProps } from "react";
import classNames from "classnames";

import type { ClassName } from "types";

export type Props = {
  /**
   * Optional class(es) to pass to the wrapping span element.
   */
  className?: ClassName;
  /**
   * Whether the spinner should have a light appearance.
   */
  isLight?: boolean;
  /**
   * Text to display next to the spinner.
   */
  text?: string;
  /**
   * What the role of the spinner should be.
   */
  role?: string;
  /**
   * The politeness setting of the spinner alert.
   */
  ariaLive?: "assertive" | "off" | "polite";
} & HTMLProps<HTMLSpanElement>;

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Spin](https://docs.vanillaframework.io/settings/animation-settings/#spin) animation.
 */
const Spinner = ({
  className,
  text,
  isLight = false,
  ariaLive = "polite",
  role = "alert",
  ...props
}: Props): JSX.Element => (
  <span
    {...props}
    className={classNames(className, "p-text--default")}
    role={role}
    aria-live={ariaLive}
  >
    <i
      className={classNames("p-icon--spinner", "u-animation--spin", {
        "is-light": isLight,
      })}
    >
      {text ? "" : "Loading"}
    </i>
    {text && (
      <>
        &ensp;
        <span>{text}</span>
      </>
    )}
  </span>
);

export default Spinner;
