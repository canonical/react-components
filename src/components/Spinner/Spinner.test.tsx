import { render, screen } from "@testing-library/react";
import React from "react";

import Spinner from "./Spinner";

describe("<Spinner />", () => {
  // snapshot tests
  it("renders and matches the snapshot", () => {
    render(<Spinner />);

    expect(screen.getByRole("alert")).toMatchSnapshot();
  });

  it("renders and matches the snapshot with text", () => {
    render(<Spinner text="foo" />);

    expect(screen.getByRole("alert")).toMatchSnapshot();
  });

  // unit tests
  it("renders a spinner icon", () => {
    render(<Spinner />);
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(".p-icon--spinner")).toBeInTheDocument();
    expect(screen.getByText("Loading")).toBeInTheDocument();
  });

  it("renders text correctly if given text prop", () => {
    const text = "Loading...";
    render(<Spinner text={text} />);

    expect(screen.getByText(text)).toBeInTheDocument();
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-icon--spinner")?.textContent
    ).not.toContain(text);
  });

  it("renders Loading... for icon text if no text prop is provided", () => {
    render(<Spinner />);
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-icon--spinner")?.textContent
    ).toBe("Loading");
  });

  it("renders an assertive spinner", () => {
    render(<Spinner ariaLive={"assertive"} />);

    expect(screen.getByRole("alert")).toHaveAttribute("aria-live", "assertive");
  });

  it("renders a light spinner if given isLight prop", () => {
    render(<Spinner isLight />);

    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-icon--spinner")
    ).toHaveClass("is-light");
  });

  it("renders a custom aria-label", () => {
    render(<Spinner aria-label="custom loading text" />);

    expect(screen.getByRole("alert")).toHaveAttribute(
      "aria-label",
      "custom loading text"
    );
  });
});
