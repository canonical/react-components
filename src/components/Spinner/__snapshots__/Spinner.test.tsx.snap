// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> renders and matches the snapshot 1`] = `
<span
  aria-live="polite"
  class="p-text--default"
  role="alert"
>
  <i
    class="p-icon--spinner u-animation--spin"
  >
    Loading
  </i>
</span>
`;

exports[`<Spinner /> renders and matches the snapshot with text 1`] = `
<span
  aria-live="polite"
  class="p-text--default"
  role="alert"
>
  <i
    class="p-icon--spinner u-animation--spin"
  />
  â€‚
  <span>
    foo
  </span>
</span>
`;
